turbo/snapshotsync/snapshotsync.go

```diff
@@ -66,11 +70,190 @@ func RequestSnapshotsDownload(ctx context.Context, downloadRequest []services.Do
+			// parse "from" (000000) and "to" (000100) from the name. 100 is 100'000 blocks
+			minusSplit := strings.Split(name, "-")
+			// convert the range to uint64
+			from, err = strconv.ParseUint(minusSplit[1], 10, 64)
+			if err != nil {
+				return nil, err
+			}
+			to, err = strconv.ParseUint(minusSplit[2], 10, 64)
+			if err != nil {
+				return nil, err
+			}
+			from *= 1000
+			to *= 1000
+			kind = minusSplit[3]
+		}
+		blackList[p.Name] = struct{}{} // Add all of them to the blacklist and remove the ones that are not blacklisted later.
+		snapshotKindToNames[kind] = append(snapshotKindToNames[kind], snapshotFileData{
+			from:      from,
+			to:        to,
+			stepBased: shouldUseStepsForPruning(name),
+			name:      name,
+		})
+	}
+	// sort the snapshots by "from" and "to" in ascending order
+	for _, snapshots := range snapshotKindToNames {
+		prunedDistance := uint64(0) // keep track of pruned distance for snapshots
+		// sort the snapshots by "from" and "to" in descending order
+		sort.Slice(snapshots, func(i, j int) bool {
+			if snapshots[i].from == snapshots[j].from {
+				return snapshots[i].to > snapshots[j].to
+			}
+			return snapshots[i].from > snapshots[j].from
+		})
+		for _, snapshot := range snapshots {
+			if snapshot.stepBased {
+				if prunedDistance >= uint64(stepPrune) {
+					break
+				}
+			} else if prunedDistance >= uint64(blockPrune) {
+				break
+			}
+
+			if snapshot.stepBased {
+				delete(blackList, snapshot.name)
+				prunedDistance += snapshot.to - snapshot.from
+				continue
+			}
+			delete(blackList, snapshot.name)
+			prunedDistance += snapshot.to - snapshot.from
+		}
+	}
+	return blackList, nil
+}
+
+// getMinimumBlocksToDownload - get the minimum number of blocks to download
+func getMinimumBlocksToDownload(tx kv.Tx, blockReader services.FullBlockReader, minStep uint64, expectedPruneBlockAmount uint64) (uint64, uint64, error) {
+	frozenBlocks := blockReader.Snapshots().SegmentsMax()
+	minToDownload := uint64(math.MaxUint64)
+	minStepToDownload := minStep
+	stateTxNum := minStep * config3.HistoryV3AggregationStep
+	if err := blockReader.IterateFrozenBodies(func(blockNum, baseTxNum, txAmount uint64) error {
+		if stateTxNum <= baseTxNum { // only cosnider the block if it
+			return nil
+		}
+		if blockNum == frozenBlocks-expectedPruneBlockAmount {
+			minStepToDownload = (baseTxNum / config3.HistoryV3AggregationStep) + 1
+		}
+		newMinToDownload := frozenBlocks - blockNum
+		if newMinToDownload < minToDownload {
+			minToDownload = newMinToDownload
+		}
+		return nil
+	}); err != nil {
+		return 0, 0, err
+	}
+	// return the minimum number of blocks to download and the minimum step.
+	return minToDownload, minStep - minStepToDownload, nil
+}
+
+func getMaxStepRangeInSnapshots(preverified snapcfg.Preverified) (uint64, error) {
+	maxTo := uint64(0)
+	for _, p := range preverified {
+		// take the "to" from "domain" snapshot
+		if !strings.HasPrefix(p.Name, "domain") {
```

Comment:
same, `ParseFileName(dir, fileName string) (res FileInfo, isE3Seedable bool, ok bool)` may be used instead

Reply:
again state files not supported by that function :(

Reply:
@Giulio2002 see 2nd return parameter.